Fix async scroll event propagation for GTK4

moggiesir opened the pull request:
#### dbc31d00f78f65a47f81c481e2eea008d13ef588
<pre>
Fix async scroll event propagation for GTK4
<a href="https://bugs.webkit.org/show_bug.cgi?id=243924">https://bugs.webkit.org/show_bug.cgi?id=243924</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=245616">https://bugs.webkit.org/show_bug.cgi?id=245616</a>

Reviewed by Michael Catanzaro.

Since GTK offers no way to asynchronously decide on whether an event should be
propagated or stopped and webkit doesn&apos;t want to block the UiProcess on event
handing in the WebProcess, webkit2gtk needs to use a bit of a hacky workaround.
This involves returning GDK_EVENT_STOP to the original event handling, sending
the event to the WebProcess for handling, which then asynchronously determines
whether the event was handled. If not, the webkit scroll handler is informed
that the &quot;next&quot; scroll event should be propagated instead of handled+stopped,
then re-emits the event for handling by GTK&apos;s run loop.

This works fine in GTK3 because there&apos;s an API to immediately run the event
loop for a given event (ensuring that the next event that the handler sees is
the intended one). For GTK4, the API for re-inserting an event appends the event
to the end of the event queue (despite what the documentation says), meaning the
next event the handler sees may not be the intended event - thus the wrong event
gets propagated and the just re-added event gets sent to the WebProcess again.

This coupled with the fact that GDK4 will synthesize scroll-start/scroll-end
events in some cases to ensure scroll events allows have the scroll-start,
scroll, scroll-end sequence can easily cause loops since it&apos;s not just 1 event
that needs to be propagated, it&apos;s up to 3.

To work better with the APIs that exist for GTK4, it seems like the best option
is to track the set of events that are &quot;pending propagation&quot; instead of relying
on the tight interlock of just ignoring the &quot;next&quot; event. The key handling
potentially also needs to be addressed similarly. Unfortunately, even tracking
the set of events to ignore is challenging because GTK4 can compress/coalesce
scroll events, meaning there will inevitably be some cases where we won&apos;t
propagate events that otherwise maybe should have been propagated (and
&quot;unhandled&quot; scroll events may cause some faster/slower than expected scrolling
when they get compressed into an not-yet-handled events).

Canonical link: <a href="https://commits.webkit.org/259530@main">https://commits.webkit.org/259530@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35c4094485c8a3bc0244b832e2f8974524464db6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105165 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114424 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174602 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15392 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5169 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97479 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110921 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11893 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94891 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39394 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93760 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81060 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94989 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7579 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27885 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/93049 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/5311 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7674 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4461 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30308 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13726 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47440 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101749 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6562 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9462 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25384 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
```diff
@@ -477,10 +477,11 @@ void PageClientImpl::wheelEventWasNotHandledByWebCore(const NativeWebWheelEvent&
 
 #if USE(GTK4)
-    gdk_display_put_event(gtk_widget_get_display(m_viewWidget), event.nativeEvent());
+    // Note: the docs for gdk_display_put_event lie - this adds to the end of the queue, not the front.
```
We need to make sure there is a GTK issue report.

moggiesir replied:
Didn't see one, so I filed https://gitlab.gnome.org/GNOME/gtk/-/issues/5525

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -270,7 +272,7 @@ struct _WebKitWebViewBasePrivate {
     IntSize viewSize { };
     bool shouldForwardNextKeyEvent { false };
-    bool shouldForwardNextWheelEvent { false };
+    Vector<GdkEvent*> shouldForwardNextWheelEvent;
```
This needs to be renamed now, e.g. `wheelEventsToForward`.

Also, if it's meant to hold ownership, then it should be `Vector<GRefPtr<GdkEvent>>`. But if it's not meant to hold ownership, then what you have here is fine.

moggiesir replied:
Done

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
```diff
@@ -52,7 +52,7 @@ void webkitWebViewBaseSetTooltipText(WebKitWebViewBase*, const char*);
 void webkitWebViewBaseSetMouseIsOverScrollbar(WebKitWebViewBase*, WebKit::WebHitTestResultData::IsScrollbar);
 void webkitWebViewBaseForwardNextKeyEvent(WebKitWebViewBase*);
-void webkitWebViewBaseForwardNextWheelEvent(WebKitWebViewBase*);
+void webkitWebViewBaseForwardNextWheelEvent(WebKitWebViewBase*, GdkEvent*);
```
This should probably be renamed to `webkitWebViewBaseForwardWheelEvent`

moggiesir replied:
Done

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
```diff
@@ -477,10 +477,11 @@ void PageClientImpl::wheelEventWasNotHandledByWebCore(const NativeWebWheelEvent&
 #if USE(GTK4)
-    gdk_display_put_event(gtk_widget_get_display(m_viewWidget), event.nativeEvent());
+    // Note: the docs for gdk_display_put_event lie - this adds to the end of the queue, not the front.
+    gdk_display_put_event(gtk_widget_get_display(m_viewWidget), gdk_event_ref(event.nativeEvent()));
```
Why the gdk_event_ref()? This call does not transfer ownership, so why is this necessary and where is the paired unref?

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1339,7 +1343,33 @@ static gboolean handleScroll(WebKitWebViewBase* webViewBase, double deltaX, doub
+        // https://docs.gtk.org/gtk4/migrating-3to4.html#adapt-to-gdkevent-api-changes
+        // There's no way to both propagate and stop the current event, so just clear out these obsolete events from the to-be-propagated vector.
+        unsigned length = 0;
+        auto* history = gdk_event_get_history(event, &length);
```
Use GUnqiuePtr, something like:

```suggestion
        GUniquePtr<GdkTimeCoord*> history(gdk_event_get_history(event, &length));
```

moggiesir replied:
Done

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1299,8 +1301,10 @@ static gboolean webkitWebViewBaseScrollEvent(GtkWidget* widget, GdkEventScroll*
 
-    if (std::exchange(priv->shouldForwardNextWheelEvent, false))
+    if (!priv->shouldForwardNextWheelEvent.isEmpty()) {
+        priv->shouldForwardNextWheelEvent.remove(0);
```
Elsewhere you unref the events before removing them from this vector, but here you don't. Need to be consistent. Does the vector hold ownership or not?

moggiesir replied:
Removed ref/unref code

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -2390,9 +2409,9 @@ void webkitWebViewBaseForwardNextKeyEvent(WebKitWebViewBase* webkitWebViewBase)
+void webkitWebViewBaseForwardNextWheelEvent(WebKitWebViewBase* webkitWebViewBase, GdkEvent* event)
 {
-    webkitWebViewBase->priv->shouldForwardNextWheelEvent = true;
+    webkitWebViewBase->priv->shouldForwardNextWheelEvent.append(event);
```
So are you taking ownership here or not? It really doesn't look like it. Notably, there is no gdk_event_ref(). But elsewhere you unref events after removing them from this vector.

Notably, we do not hold or transfer ownership via a raw pointer. Use `GRefPtr<GdkEvent>` to hold ownership or `GRefPtr<GdkEvent>&&` to transfer ownership. `GdkEvent*` is fine when not moving ownership.

moggiesir replied:
I don't need to take ownership, but I thought ref/unref might be more clear. Using GRefPtr would have been a bit messy to set up because of the difference between gtk3 and gtk4 (gdk_event_copy/g_event_free vs gdk_event_ref/gdk_event_unref).

Anyways, dropped all the ownership since it isn't needed.

mcatanzaro reviewed (commented):
Thanks for this significant pull request. I've not quite wrapped my head around the event handling code, so we'll want to wait for Carlos Garcia to take a look at this. But I do have some preliminary comments.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBasePrivate.h:
```diff
@@ -52,7 +52,7 @@ void webkitWebViewBaseSetTooltipText(WebKitWebViewBase*, const char*);
 void webkitWebViewBaseSetTooltipArea(WebKitWebViewBase*, const WebCore::IntRect&);
 void webkitWebViewBaseSetMouseIsOverScrollbar(WebKitWebViewBase*, WebKit::WebHitTestResultData::IsScrollbar);
 void webkitWebViewBaseForwardNextKeyEvent(WebKitWebViewBase*);
```
Would the same strategy suffice to fix key events too?

moggiesir replied:
Sure, added (and it's simpler because key events shouldn't get compressed/coalesced)

alice-mkh commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1350,6 +1389,7 @@ static gboolean handleScroll(WebKitWebViewBase* webViewBase, double deltaX, doub
         // We only have hold events on touchpads, so assume one if we get no event.
         GdkInputSource source = device ? gdk_device_get_source(device) : GDK_SOURCE_TOUCHPAD;
 
+        bool isEnd = gdk_event_is_scroll_stop_event(event);
```
Wait what? This event can be null, what happens then?

moggiesir replied:
Changed to handle event being null.

FWIW, event shouldn't be null here AFAICT and below it seems to assume the event is not null (calls gdk_event_get_modifier_state).

alice-mkh commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1350,12 +1389,13 @@ static gboolean handleScroll(WebKitWebViewBase* webViewBase, double deltaX, doub
     }
 
-    if (!isEnd && shouldInvertDirectionForScrollEvent(priv->mouseIsOverScrollbar, gdk_event_get_modifier_state(event) & GDK_SHIFT_MASK))
+    if (shouldInvertDirectionForScrollEvent(priv->mouseIsOverScrollbar, gdk_event_get_modifier_state(event) & GDK_SHIFT_MASK))
```
This use? Well, one thing with null events is that can only happen for scroll-end (that signal emission is synthesized if you have old libinput without hold gestures support/on trackpoint/possible on semi-mt touchpads, dunno about those.

So the !isEnd check here was accounting for that. :)

moggiesir replied:
Sure, I added the !isEnd check back here. I don't know these libinput edge cases, so feel free to be prescriptive about how to handle them. e.g. do I need to bring back the scroll-end listener for this?

alice-mkh replied:
Oh, missed that. Yes, you absolutely need both that and scroll-begin, otherwise it's going to be very very broken on touchpads.

- scroll-end is needed for kinetic scrolling
- both of them are needed for hold gestures

moggiesir replied:
Done

alice-mkh commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -1350,12 +1389,13 @@ static gboolean handleScroll(WebKitWebViewBase* webViewBase, double deltaX, doub
         // We only have hold events on touchpads, so assume one if we get no event.
         GdkInputSource source = device ? gdk_device_get_source(device) : GDK_SOURCE_TOUCHPAD;
 
+        bool isEnd = event ? gdk_event_is_scroll_stop_event(event) : false;
```
And since null events only happen for scroll-end, `false` is absolutely wrong here.

moggiesir replied:
Changed to true as the fallback then.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -968,6 +975,43 @@ static void webkitWebViewBaseUnmap(GtkWidget* widget)
+        // https://docs.gtk.org/gtk4/migrating-3to4.html#adapt-to-gdkevent-api-changes
+        // There's no way to both propagate and stop the current event, so just clear out these obsolete events from the to-be-propagated vector.
+        unsigned length = 0;
+        GUniquePtr<GdkTimeCoord> history(gdk_event_get_history(event, &length));
```
Huh, this is the best way to write it currently, but it's very confusing and would fail if GdkTimeCoord were to ever gain a more specialized deleter in the future. That's very unlikely to happen in GTK 4, but who knows what the future holds. Problem is it looks like you have a pointer to delete a GdkTimeCoord, but what you actually have is a pointer to delete an array of unowned GdkTimeCoords.

I don't want to ask you to spend much time on this, but please do a quick/easy test to see if `GUniquePtr<GdkTimeCoord, GFreeDeleter>` would work. You would need to add `GFreeDeleter` in GUniquePtr.h. It would be exactly the same as GPtrDelete except there would be no specializations for it, so you can clearly guarantee what the deleter func is. This is optional; if you have any trouble doing so, then don't worry about it.

moggiesir replied:
Is this what you were thinking?

mcatanzaro replied:
Yes, perfect.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -2385,14 +2432,28 @@ void webkitWebViewBaseDidPerformDragControllerAction(WebKitWebViewBase* webViewB
-    webkitWebViewBase->priv->shouldForwardNextKeyEvent = TRUE;
+#if USE(GTK4)
+    webkitWebViewBase->priv->keyEventsToPropagate.append(event);
+    // Note: the docs for gdk_display_put_event lie - this adds to the end of the queue, not the front.
```
Please add the GTK bug report link to this comment.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -968,6 +975,43 @@ static void webkitWebViewBaseUnmap(GtkWidget* widget)
+                    auto isMatch = gdk_event_get_time(current.get()) == oldTime;
+                    return isMatch;
+                });
+            }
```
I don't understand what's happening here. The comment makes it look like you're trying to remove the last history entry for each event, but in fact your code is looking at all history entries and removing the first event that matches the time of the history entry. Is that intended? Why?

moggiesir replied:
Changed the code a bit to try to make it more clear the comment just pertains to the loop end condition adjustment.

The comment describing what this whole block is trying to do is a few lines above:
```
// Some events that were meant for delayed propagation may have been coalesced/compressed by GDK into the current event:
// https://docs.gtk.org/gtk4/migrating-3to4.html#adapt-to-gdkevent-api-changes
// There's no way to both propagate and stop the current event, so just clear out these obsolete events from the to-be-propagated vector.
```

So sometimes PageClientImpl tries to "propagate" an event after-the-fact, but when it adds the event to the GDK event queue, GDK ends up coalescing the event with some other active scroll event, meaning we'll never see the event we want to propagate in handleScroll(). So this code detects if an event that was meant for propagation was coalesced with the current event, and if so cleans up wheelEventsToPropagate so that we don't leak memory.

mcatanzaro reviewed (commented):
Dunno how correct this it, but it looks plausible and seems better than what we have now, so I'm inclined to approve it, especially now that Exalm has reviewed it.

But I also have some more thoughts.

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -968,6 +975,44 @@ static void webkitWebViewBaseUnmap(GtkWidget* widget)
+            for (unsigned i = 0; i < length; i++) {
+                auto oldTime = history.get()[i].time;
+                webViewBase->priv->wheelEventsToPropagate.removeFirstMatching([&oldTime] (GRefPtr<GdkEvent>& current) {
+                    return gdk_event_get_time(current.get()) == oldTime;
```
Is it impossible to have more than one such matching event?

moggiesir replied:
Not sure, so changed to removeAllMatching

moggiesir committed:
Fix async scroll event propagation for GTK4
https://bugs.webkit.org/show_bug.cgi?id=243924
https://bugs.webkit.org/show_bug.cgi?id=245616

Reviewed by Michael Catanzaro.

Since GTK offers no way to asynchronously decide on whether an event should be
propagated or stopped and webkit doesn't want to block the UiProcess on event
handing in the WebProcess, webkit2gtk needs to use a bit of a hacky workaround.
This involves returning GDK_EVENT_STOP to the original event handling, sending
the event to the WebProcess for handling, which then asynchronously determines
whether the event was handled. If not, the webkit scroll handler is informed
that the "next" scroll event should be propagated instead of handled+stopped,
then re-emits the event for handling by GTK's run loop.

This works fine in GTK3 because there's an API to immediately run the event
loop for a given event (ensuring that the next event that the handler sees is
the intended one). For GTK4, the API for re-inserting an event appends the event
to the end of the event queue (despite what the documentation says), meaning the
next event the handler sees may not be the intended event - thus the wrong event
gets propagated and the just re-added event gets sent to the WebProcess again.

This coupled with the fact that GDK4 will synthesize scroll-start/scroll-end
events in some cases to ensure scroll events allows have the scroll-start,
scroll, scroll-end sequence can easily cause loops since it's not just 1 event
that needs to be propagated, it's up to 3.

To work better with the APIs that exist for GTK4, it seems like the best option
is to track the set of events that are "pending propagation" instead of relying
on the tight interlock of just ignoring the "next" event. The key handling
potentially also needs to be addressed similarly. Unfortunately, even tracking
the set of events to ignore is challenging because GTK4 can compress/coalesce
scroll events, meaning there will inevitably be some cases where we won't
propagate events that otherwise maybe should have been propagated (and
"unhandled" scroll events may cause some faster/slower than expected scrolling
when they get compressed into an not-yet-handled events).

Canonical link: https://commits.webkit.org/259530@main
